custom/icds/location_reassignment/download.py

```diff
@@ -1,10 +1,129 @@
 import io
+import itertools
+from collections import defaultdict
+from copy import deepcopy
+
+from memoized import memoized
+from openpyxl import Workbook
+
+from corehq.apps.es import UserES
+from corehq.apps.locations.models import SQLLocation
 
 
 class Download(object):
-    def __init__(self, domain, parent_location_id, leaf_location_type):
-        pass
+    def __init__(self, location_id):
+        """
+        Generates an Excel file stream
+        With details of all locations related to a location
+        and users assigned to these locations.
+        Each sheet corresponds to a location type.
+        """
+        self.location = SQLLocation.active_objects.get(location_id=location_id)
 
     def dump(self):
+        self._init_location_details()
+        self._populate_assigned_users()
+        wb = self._create_workbook()
         stream = io.BytesIO()
+        wb.save(stream)
         return stream
+
+    def _init_location_details(self):
+        # setup details for the main location and all related locations
+        self._location_details_by_location_id = {}
+        for location in self._locations():
+            self._location_details_by_location_id[location.location_id] = {
+                'name': location.name,
+                'site_code': location.site_code,
+                'lgd_code': location.metadata.get('lgd_code', ''),
+                'parent_location_id': location.parent_location_id,
+                'location_type': location.location_type.code,
+                'assigned_users': [],
+            }
+        return self._location_details_by_location_id
+
+    def _locations(self):
+        # fetch all locations necessary for this download request
+        ancestors = list(self.location.get_ancestors().select_related('location_type'))
+        self_and_descendants = list(self.location.get_descendants(include_self=True)
+                                    .select_related('location_type'))
+        return ancestors + self_and_descendants
+
+    def _populate_assigned_users(self):
+        # allot assign user details to each location
+        for user_detail in self._get_users():
+            username, assigned_location_ids = user_detail
+            if not isinstance(assigned_location_ids, list):
+                assigned_location_ids = [assigned_location_ids]
```

Comment:
`get_users` just pulls the users from ES, why should this be moved there?

Reply:
Because it would make `get_users` return a more consistent value, where all `assigned_location_ids` are lists. As is, any other caller of `get_users` would need to do this same check.

Reply:
Aha, i see.

https://github.com/dimagi/commcare-hq/pull/26951/commits/51f3732559a99a4191e2aa37dc34b61186043f44